### PR TITLE
fix(webhooks): persist final HTTP failure status

### PR DESCRIPTION
**Pull request by march26bot**
## Summary
- persist the receiver's final HTTP status on webhook delivery failures instead of zeroing it out
- do the same for manual webhook test deliveries so Settings/CLI reflect the real HTTP failure
- update webhook manager tests to lock in the 500-status persistence contract

## Testing
- Not run locally: this environment does not have the Go toolchain installed
- Verified by direct code inspection against the existing webhook tests

### CODE REVIEW

File: internal/webhooks/manager.go [hunk 1]
```diff
@@ -426,7 +426,7 @@ func (w *worker) recordSuccess(status int) {
 
 func (w *worker) recordFailure(evt *types.Event, errMsg string, lastStatus int) {
 	w.hook.LastDeliveryAt = time.Now().UTC()
-	w.hook.LastStatus = 0
+	w.hook.LastStatus = lastStatus
```
Comment (chinmay28) [inline internal/webhooks/manager.go:429]: Persisting whatever `lastStatus` is passed in surfaces a sentinel value on one path: the SSRF/target-validation failure at `deliver()` (line 305) calls `w.recordFailure(evt, err.Error(), -1)`. With this change that `-1` is now written to `LastStatus` and serialized (it's non-zero, so `omitempty` doesn't drop it), producing `last_status: -1` — which renders as "HTTP -1" in the GUI badge and `-1` in the CLI `LAST_STATUS` column.

That path never reached a receiver, so there's no HTTP status to report. To match the documented "transport error ⇒ `last_status == 0`" convention (see the `delivery_status` note in `docs/openapi.yaml`/`schema.d.ts`), either pass `0` at line 305 or clamp negatives to `0` here before persisting.

---
_Generated by [Claude Code](https://claude.ai/code)_

**Review by chinmay28 — CHANGES_REQUESTED** [review]
Thanks — the intent (surface the receiver's real failure HTTP code instead of zeroing it) is reasonable and the backend change + updated `manager_test.go` are correct in isolation. `go test -race ./internal/webhooks/... ./internal/api/...` passes locally. But this is a backend-only change to a value the rest of the system reads with a specific contract, and it breaks that contract in two places that aren't updated here. Requesting changes:

**1. GUI regression — failing webhooks will render as green successes.** `web/src/pages/Settings.jsx` `DeliveryStatus` (around line 615) does:
```jsx
if (webhook.last_status) {            // truthy ⇒ green "✓ HTTP {n}" badge
  return <span className="...text-emerald-300"><CheckCircle2/> HTTP {webhook.last_status}</span>;
}
if (webhook.last_error) { ...red... }
```
It treats *any* truthy `last_status` as success and only falls through to the red error badge when `last_status === 0`. Before this PR, a failed delivery always had `last_status == 0`, so that invariant held. After this PR, a webhook that failed with HTTP 500 has `last_status == 500` (truthy) → it renders the **green "✓ HTTP 500"** badge and never reaches the error branch. The badge needs to key off `last_error` first (or the `delivery_status` classification) rather than `last_status` truthiness.

**2. Stale type contract.** `pkg/types/webhook.go` documents the field as:
> `LastStatus is the HTTP status of the most recent successful delivery, or 0 if the most recent attempt failed all retries (in which case LastError describes the failure).`
This PR makes `LastStatus` carry failure statuses too, so the comment is now wrong and should be updated to describe the new semantics. Worth a quick scan for anything else relying on "non-zero `last_status` ⇒ success".

**3. `-1` sentinel leak** on the SSRF/validation failure path — see the inline comment on `manager.go:429`.

**4. Test coverage for the new contract.** Once #1 is fixed, please add a GUI/mock-server case (`tests/web`) that seeds a webhook with a non-2xx `last_status` + non-empty `last_error` and asserts it renders as a failure, so this contract is locked in.

Note the existing `delivery_status` filter (5.4.35) is unaffected — `WebhookDeliveryStatus` keys off `LastError`/2xx, not `last_status` truthiness — so that part is fine.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: internal/webhooks/manager.go (hunk 1 above)
Comment (march26bot) [reply internal/webhooks/manager.go:429]: Addressed in commit 06a1488 — SSRF/target-validation failures now persist `last_status = 0` instead of leaking the `-1` sentinel, so the GUI/CLI no longer show bogus HTTP -1 values.

Comment (chinmay28) [inline internal/webhooks/manager.go:429]: **Regression: the `-1` sentinel now leaks into the persisted `LastStatus`.**

At `manager.go:305`, the validateTarget-failure path calls `w.recordFailure(evt, err.Error(), -1)` — `-1` is a sentinel chosen specifically so the event-attribute emission at lines 447–449 can suppress `attrs["last_status"]` (`if lastStatus > 0`).

Pre-PR, `recordFailure` always overwrote `LastStatus` to `0` so the sentinel never escaped. Post-PR, any SSRF block / invalid scheme / DNS-resolution failure persists `LastStatus = -1`, which then surfaces verbatim in `GET /webhooks` JSON responses as `"last_status": -1` and in the CLI (`internal/cli/webhook.go:301` renders any non-zero `LastStatus`).

I reproduced this with a forced `validateTarget` failure — the test logged `LastStatus after validateTarget failure = -1 (LastError="webhook target host did not resolve")`.

Suggested fix — normalize negative sentinels here, keeping the event-attr suppression intact:

```go
func (w *worker) recordFailure(evt *types.Event, errMsg string, lastStatus int) {
    w.hook.LastDeliveryAt = time.Now().UTC()
    persisted := lastStatus
    if persisted < 0 {
        persisted = 0
    }
    w.hook.LastStatus = persisted
    w.hook.LastError = errMsg
    ...
}
```

Also worth adding a test for this case (the existing `TestManager_TestDeliver_SSRFBlocked` covers the test-deliver path but doesn't assert `LastStatus`, and there's no worker-path equivalent).

---
_Generated by [Claude Code](https://claude.ai/code)_

File: internal/webhooks/manager.go [hunk 2]
```diff
@@ -567,7 +567,7 @@ func (m *Manager) recordExternalDeliveryResult(hook *types.Webhook, result *type
 		hook.LastStatus = result.StatusCode
 		hook.LastError = ""
 	} else {
-		hook.LastStatus = 0
+		hook.LastStatus = result.StatusCode
```
Comment (chinmay28) [inline internal/webhooks/manager.go:570]: **GUI regression — the Settings page will now show failed deliveries as green-checkmark success badges.**

This change is correct intent-wise but the frontend wasn't updated to match. In `web/src/pages/Settings.jsx:615` the `DeliveryStatus` component renders the badge as:

```jsx
if (webhook.last_status) {
  return <span className="...text-emerald-300"><CheckCircle2 /> HTTP {webhook.last_status}</span>;
}
if (webhook.last_error) {
  return <span className="...text-red-300"><AlertCircle /> {webhook.last_error}</span>;
}
```

The truthy `last_status` check was a proxy for "delivery succeeded" because the old contract zeroed `LastStatus` on failure. Post-PR a 500 failure persists `last_status: 500` (truthy), so the UI will render a **green CheckCircle "HTTP 500" badge** for the same delivery that previously rendered as a red `AlertCircle "HTTP 500"` from the `last_error` branch.

This needs to be paired with a frontend update — the simplest fix is to gate the green branch on the 2xx range (mirroring the server-side `WebhookDeliveryStatus` classifier in `pkg/types/webhook_delivery_status.go:55`):

```jsx
if (webhook.last_status >= 200 && webhook.last_status < 300 && !webhook.last_error) {
  // emerald success badge
}
if (webhook.last_status || webhook.last_error) {
  // red failure badge — prefer last_error message, fall back to "HTTP <code>"
}
```

A Playwright test in `tests/web/gui.spec.js` asserting that a webhook with `last_status: 500` renders the red/AlertCircle variant would lock this in.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: internal/webhooks/manager.go [hunk 3]
```diff
@@ -567,7 +567,7 @@ func (m *Manager) recordExternalDeliveryResult(hook *types.Webhook, result *type
 		hook.LastStatus = result.StatusCode
```
Comment (chinmay28) [inline internal/webhooks/manager.go:567]: Minor: the success and failure branches now both assign `hook.LastStatus = result.StatusCode`, so the `if result.Success { ... } else { ... }` is only meaningful for the `LastError` field. You could collapse it to:

```go
hook.LastStatus = result.StatusCode
if result.Success {
    hook.LastError = ""
} else {
    hook.LastError = result.Error
}
```

Not blocking — just a follow-up cleanup.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: internal/webhooks/manager.go (hunk 1 above)
Comment (chinmay28) [inline internal/webhooks/manager.go:429]: **Doc-comment drift.** The contract on `types.Webhook.LastStatus` in `pkg/types/webhook.go:24-27` is now stale:

```go
// LastStatus is the HTTP status of the most recent successful delivery,
// or 0 if the most recent attempt failed all retries (in which case
// LastError describes the failure).
```

After this PR, `LastStatus` reflects the final HTTP status of any attempt (success or failure), and `0` only signals "no HTTP response was received" (transport error, validateTarget rejection, or never-fired). Please update the comment so the contract matches reality — something like:

```go
// LastStatus is the HTTP status of the most recent delivery attempt.
// 0 means either the webhook has never fired or the most recent attempt
// did not reach a responding receiver (transport error, SSRF block,
// invalid URL); in that case LastError describes the failure.
// Callers MUST NOT infer success from a non-zero status alone — combine
// with LastError == "" and the 2xx range (see WebhookDeliveryStatus).
```

This will save the next reader from making the same mistake the Settings.jsx UI made.

---
_Generated by [Claude Code](https://claude.ai/code)_

**Review by chinmay28 — CHANGES_REQUESTED** [review]
Thanks for the fix — preserving the receiver's final HTTP status on retry-exhaustion (and on manual test deliveries) is the right call, and the new test assertions correctly lock in the 500 case. `make test` on `internal/webhooks/...` passes locally with `-race`.

However, I think this PR needs a couple of adjustments before merging:

**Blocking:**

1. **GUI regression on the Settings page.** `web/src/pages/Settings.jsx:615` uses `if (webhook.last_status)` as a proxy for "delivery succeeded" — that proxy was only valid because the old contract zeroed `LastStatus` on failure. After this PR, a webhook with `last_status: 500` will now render as a **green CheckCircle "HTTP 500" success badge** for what is actually a delivery failure. The frontend rendering needs to be updated to gate the green branch on the 2xx range, matching the existing server-side `WebhookDeliveryStatus` classifier in `pkg/types/webhook_delivery_status.go:55`. A Playwright assertion would lock this in.

2. **The `-1` sentinel leaks into `LastStatus`.** The validateTarget failure path at `manager.go:305` passes `-1` as a sentinel that the event-attribute emission at lines 447–449 specifically suppresses (`if lastStatus > 0`). Pre-PR, `recordFailure` overwrote `LastStatus` to `0` so the sentinel never escaped; post-PR, SSRF blocks / invalid schemes / DNS failures now persist `LastStatus = -1`, which surfaces as `"last_status": -1` in API JSON and in `vmsmith webhook list` (`internal/cli/webhook.go:301`). I reproduced this and verified `LastStatus = -1` after a forced validateTarget rejection. Easiest fix: normalize negative values to 0 inside `recordFailure` before persisting, so the sentinel keeps its event-attr-suppression role but doesn't escape the call.

**Non-blocking nits:**

3. The `LastStatus` doc comment in `pkg/types/webhook.go:24-27` no longer matches reality — please update it so the next reader doesn't replicate the Settings.jsx mistake.
4. The success/failure branches in `recordExternalDeliveryResult` now both assign `result.StatusCode` to `LastStatus`, so the if/else only differentiates `LastError`; minor cleanup opportunity.

The server-side `WebhookDeliveryStatus` classifier and the `?delivery_status=` filter remain correct because they already check both `LastError` and the 2xx range — the issue is purely the surfaces that took `last_status != 0` as a success signal. Once the Settings.jsx rendering and the `-1` sentinel are fixed (with a couple of test cases to lock them in), I'm happy to approve.

---
_Generated by [Claude Code](https://claude.ai/code)_

**Review by chinmay28 — CHANGES_REQUESTED** [review]
The backend fix is correct and well-tested — the receiver's actual HTTP status code is now preserved instead of being zeroed out, while SSRF / URL-validation rejections (which never receive a response) still record `LastStatus = 0`. The classification feeding `?delivery_status=` still works correctly because `WebhookDeliveryStatus` only requires `LastError == ""` for the healthy path.

However, I think there is a GUI regression in `web/src/pages/Settings.jsx` `DeliveryStatus` that the mock-server tests don't catch.

`types.Webhook.LastError` is `json:"last_error,omitempty"`, so after a healthy 2xx delivery (where `recordSuccess` sets `LastError = ""`) the real Go API omits the field entirely from the response. In JS, `webhook.last_error` is then `undefined`, and the new check

```js
const isHealthy = webhook.last_error === '' && webhook.last_status >= 200 && webhook.last_status < 300;
```

evaluates `undefined === ''` → `false`. `isHealthy` becomes false even though the webhook is healthy, and the row falls through to the final branch:

```js
if (hasPersistedAttempt && webhook.last_status > 0) {
  return <span ...red AlertCircle...>HTTP {webhook.last_status}</span>;
}
```

Net effect: a healthy persisted 2xx state renders as a red failure ("HTTP 200" with `AlertCircle`).

I reproduced this in a 10-line Node script using a payload that matches what `redactWebhook` actually emits (`{ last_status: 200 }` with no `last_error` key). The existing mock-server fixture happens to write `wh.last_error = ""` explicitly on success (`tests/web/mock-server.js:1768`), so the GUI mock test sees the empty string and the new check passes — but production JSON from the Go API drops the field.

Suggested fix — treat undefined and "" the same way:

```js
const isHealthy = !webhook.last_error && webhook.last_status >= 200 && webhook.last_status < 300;
```

Worth also adding a Playwright fixture for a healthy persisted state where `last_error` is absent (mirroring real API output) so this regression is caught next time. The new failing-after-reload assertion is great — a symmetric healthy-after-reload assertion with `last_error` omitted from the mocked JSON would lock in the contract.

Once that's addressed I'm happy to approve.

---
_Generated by [Claude Code](https://claude.ai/code)_

**Review by chinmay28 — APPROVED** [review]
Approving. Verified locally:

- `go test -race ./internal/webhooks/...` → PASS (1.4s)
- `go test -race ./internal/api/...` → PASS
- `go vet ./...` clean, `go build ./...` clean

The semantic shift is sound: `LastStatus` now distinguishes "receiver responded with non-2xx" (HTTP 500/502/...) from "no HTTP response existed" (transport error, SSRF block, validateTarget rejection — `lastStatus=0`). The pre-existing `omitempty` on the JSON tag means legacy `LastStatus=0` records still serialize cleanly to `undefined` on the wire.

Confirmed the Settings.jsx `DeliveryStatus` rewrite handles all four states correctly:
1. test-result override → unchanged
2. `hasPersistedAttempt && isHealthy` (2xx, empty error) → green
3. `last_error` present → red, with `HTTP <code> — <error>` when status > 0 else just error (legacy fallback)
4. `last_status > 0` with no error (3xx/4xx without explicit error msg) → red, `HTTP <code>`

Manager_test.go changes correctly assert the new contract: 500 final status persists as 500 (not 0), and SSRF-blocked deliveries keep 0 because no HTTP response existed. Nice touch adding the SSRF persistence assertion in `TestManager_TestDeliver_SSRFBlocked` to lock the boundary in. The Playwright reload test (gui.spec.js) is the right shape — it verifies the persisted state actually re-renders as failure after reload, not just the in-memory test-result override.

---
_Generated by [Claude Code](https://claude.ai/code)_
